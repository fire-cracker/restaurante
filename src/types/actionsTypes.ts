export interface Action<T> {
  type: ActionType;
  payload?: T;
}

export enum ActionType {
  SIGNUP_REQUEST_PENDING = 'SIGNUP_REQUEST_PENDING',
  SIGNUP_REQUEST_SUCCESS = 'SIGNUP_REQUEST_SUCCESS',
  SIGNUP_REQUEST_FAILED = 'SIGNUP_REQUEST_FAILED',
  
  LOGIN_REQUEST_PENDING = 'LOGIN_REQUEST_PENDING',
  LOGIN_REQUEST_SUCCESS = 'LOGIN_REQUEST_SUCCESS',
  LOGIN_REQUEST_FAILED = 'LOGIN_REQUEST_FAILED',

  GET_PROFILE_REQUEST_PENDING = 'GET_PROFILE_REQUEST_PENDING',
  GET_PROFILE_REQUEST_SUCCESS = 'GET_PROFILE_REQUEST_SUCCESS',
  GET_PROFILE_REQUEST_FAILED = 'GET_PROFILE_REQUEST_FAILED',

  LOGOUT_REQUEST_SUCCESS = 'LOGOUT_REQUEST_SUCCESS',

  SET_USER_STATE = 'SET_USER_STATE'
}