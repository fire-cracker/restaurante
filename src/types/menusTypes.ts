export interface IMenu {
  id: number
  name: string
  image: string
  price: number
  recipe: string
  type: string
}

export interface IMenuState {
  menus: IMenu | null
  count: number
  fetching: boolean
  fetched: boolean
}
