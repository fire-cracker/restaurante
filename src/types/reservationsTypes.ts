export interface IReservation {
  id: number
  userId: string
  date: string
  time: string
  type: string
  price: number
  persons: number
}
export interface INewReservation {
  date: string
  time: string
  type: string
  persons: number
}

export interface IReservationState {
  reservation: IReservation | null
  fetching: boolean
  fetched: boolean
}

export interface IStripeCharge {
  id: string
  object: string
  amount: 2000
  paid: boolean
  receipt_email: string
  receipt_number: string
  receipt_url: string
  status: string
}
