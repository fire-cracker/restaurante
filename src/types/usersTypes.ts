export interface IUser {
  id: string
  username: string
  email: string
  role: 'customer' | 'admin'
}

export interface IUserState {
  user: IUser | null
  logingIn: boolean
  isLoggedIn: boolean
}