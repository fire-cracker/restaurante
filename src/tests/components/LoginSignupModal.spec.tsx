import React from 'react'
import { BrowserRouter } from 'react-router-dom'
import { render, cleanup } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'

import LoginSignupModal from '../../components/LoginSignupModal'
import { IUser, IUserState } from '../../types/usersTypes'

const createMockStore = configureMockStore([thunk])

interface IProps {
  show: boolean
  onHide: () => void
  login: (email: string, password: string) => Promise<IUser>
  signup: (username: string, email: string, password: string) => Promise<IUser>
  userState: IUserState
  onhandleChange: () => void
  toggleLoginSignup: () => void
}

describe('Login/Signup Modal', () => {
  const defaultProps: IProps = {
    show: true,
    onHide: jest.fn(),
    login: jest.fn(),
    signup: jest.fn(),
    userState: {
      user: null,
      logingIn: false,
      isLoggedIn: false
    },
    onhandleChange: jest.fn(),
    toggleLoginSignup: jest.fn()
  }

  const initialState = {
    userState: {
      user: null,
      logingIn: false,
      isLoggedIn: false
    }
  }

  const setup = (newProps?: any, newStore?: any) => {
    const props = { ...defaultProps, ...newProps }
    const store = createMockStore({
      ...initialState,
      userState: { ...initialState.userState, ...newStore }
    })

    return render(
      <BrowserRouter>
        <Provider store={store}>
          <LoginSignupModal {...props} />
        </Provider>
      </BrowserRouter>
    )
  }

  afterAll(() => {
    jest.clearAllMocks()
    cleanup()
  })

  test('renders Login Modal', () => {
    const wrapper = setup()
    expect(wrapper).toMatchSnapshot()
    const { getAllByText } = wrapper
    const linkElement = getAllByText('Login')
    expect(linkElement).toHaveLength(2)
  })

  test('should trigger onchange when email is changed', async () => {
    const wrapper = setup()
    const { getByPlaceholderText } = wrapper
    const emailInput = getByPlaceholderText('email')
    expect(emailInput).toBeInTheDocument()
    await userEvent.type(emailInput, 'janedoe@example.com')
    // expect(defaultProps.onhandleChange).toHaveBeenCalledTimes(16)
    expect(emailInput).toHaveValue('janedoe@example.com')
  })

  test('should trigger onchange when password is changed', async () => {
    const wrapper = setup()
    const { getByPlaceholderText } = wrapper
    const passwordInput = getByPlaceholderText('password')
    expect(passwordInput).toBeInTheDocument()
    await userEvent.type(passwordInput, 'password')
    // expect(defaultProps.onhandleChange).toHaveBeenCalledTimes(8)
    expect(passwordInput).toHaveValue('password')
  })

  // eslint-disable-next-line jest/expect-expect
  it('should login user if user provides correct login credentials', async () => {
    const wrapper = setup()
    const { getByPlaceholderText, getAllByText } = wrapper
    const emailInput = getByPlaceholderText('email')
    const passwordInput = getByPlaceholderText('password')
    await userEvent.type(emailInput, 'janedoe@example.com')
    await userEvent.type(passwordInput, 'password')
    const loginButtons = getAllByText('Login')
    await userEvent.click(loginButtons[1])
    // expect(defaultProps.login).toHaveBeenCalled()
  })

  it('should not login user if user provides incorrect login credentials', async () => {
    const wrapper = setup()
    const { getByPlaceholderText, getAllByText } = wrapper
    const emailInput = getByPlaceholderText('email')
    const passwordInput = getByPlaceholderText('password')
    await userEvent.type(emailInput, '')
    await userEvent.type(passwordInput, '')
    const loginButtons = getAllByText('Login')
    await userEvent.click(loginButtons[1])
    expect(defaultProps.login).toHaveBeenCalledTimes(0)
  })

  test('user can navigate to signup page', async () => {
    const { getByText } = setup()
    const signupLink = getByText('Sign Up')
    await userEvent.click(signupLink)
    // expect(defaultProps.toggleLoginSignup).toHaveBeenCalledTimes(1)
  })

  test('user can signup if user provides correct signup credentials', async () => {
    const { getByText, getByPlaceholderText, getAllByText } = setup({ loginState: true })
    const signupLink = getByText('Sign Up')
    await userEvent.click(signupLink)
    const usernameInput = getByPlaceholderText('username')
    await userEvent.type(usernameInput, 'janedoe')
    expect(usernameInput).toHaveValue('janedoe')
    const signupButtons = getAllByText('Signup')
    await userEvent.click(signupButtons[1])
    // expect(defaultProps.signup).toHaveBeenCalled()
  })
})
