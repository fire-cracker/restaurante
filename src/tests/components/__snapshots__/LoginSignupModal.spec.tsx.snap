// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login/Signup Modal renders Login Modal 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="modal-open"
    style="padding-right: 0px;"
  >
    <div
      aria-hidden="true"
    />
    <div
      class="fade modal-backdrop show"
    />
    <div
      aria-labelledby="contained-modal-title-vcenter"
      aria-modal="true"
      class="fade modal-wrapper modal show"
      role="dialog"
      style="display: block;"
      tabindex="-1"
    >
      <div
        class="modal-dialog modal-lg modal-dialog-centered"
        role="document"
      >
        <div
          class="modal-content"
        >
          <div
            class="wrapper-row row"
          >
            <div
              class="wrapper-row-col p-3 justify-content-center align-items-center col"
            >
              <div
                class="border-0 p-0 pr-2 pt-2 modal-header"
              >
                <button
                  class="close"
                  type="button"
                >
                  <span
                    aria-hidden="true"
                  >
                    ×
                  </span>
                  <span
                    class="sr-only"
                  >
                    Close
                  </span>
                </button>
              </div>
              <div
                class="modal-body"
              >
                <div
                  class="modal-body-row justify-content-center align-items-center text-center row"
                >
                  <div
                    class="flex-column justify-content-center align-items-center col-md-12"
                  >
                    <div
                      class="col"
                    >
                      <h2>
                        Login
                      </h2>
                    </div>
                    <div
                      class="divider mt-1 mb-3 text-darkkhaki d-flex flex-row justify-content-center align-items-center col"
                    >
                      <p
                        class="line my-2 mx-1"
                      >
                         
                      </p>
                      <p
                        class="icon"
                      >
                        <img
                          alt="entwined"
                          class="fa fa-cutlery"
                          src="/images/entwined.svg"
                        />
                      </p>
                      <p
                        class="line my-2 mx-1"
                      >
                         
                      </p>
                    </div>
                  </div>
                  <div
                    class="justify-content-center align-items-center col-md-12"
                  >
                    <form
                      class=""
                      novalidate=""
                    >
                      <div
                        class="flex-column justify-content-center align-items-center form-row"
                      >
                        <div
                          class="form-group col"
                        >
                          <label
                            class="form-label"
                            for="validationCustom01"
                          />
                          <input
                            class="form-control"
                            id="validationCustom01"
                            name="email"
                            placeholder="email"
                            required=""
                            type="text"
                          />
                          <div
                            class="valid-feedback"
                          >
                            Looks good!
                          </div>
                        </div>
                        <div
                          class="form-group col"
                        >
                          <label
                            class="form-label"
                            for="validationCustom02"
                          />
                          <input
                            class="form-control"
                            id="validationCustom02"
                            name="password"
                            placeholder="password"
                            required=""
                            type="password"
                          />
                          <div
                            class="valid-feedback"
                          >
                            Looks good!
                          </div>
                        </div>
                      </div>
                      <div
                        class="justify-content-center align-items-center form-row"
                      >
                        <button
                          class="bg-black border-0 rounded-0 btn btn-primary"
                          type="submit"
                        >
                          Login
                        </button>
                      </div>
                    </form>
                  </div>
                </div>
              </div>
              <div
                class="justify-content-center align-items-center modal-footer"
              >
                Don't have an account?
                <button
                  class="btn btn-link"
                  type="button"
                >
                  Sign Up
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
